sw/airborne/firmwares/rotorcraft/stabilization/stabilization_attitude_heli_indi.c

```diff
@@ -0,0 +1,618 @@
+  }
+
+  /* RADIO throttle stick value, for 4dof mode */
+  //int32_t accel_z_sp = (-1)*3*((guidance_v_rc_delta_t - MAX_PPRZ/2) << INT32_ACCEL_FRAC) / (MAX_PPRZ/2);
+  //accel_z_sp = ((accel_z_sp << INT32_TRIG_FRAC) / guidance_v_thrust_coeff);
+
+  /* Transform yaw into a delta yaw while keeping filtered yawrate (kinda hacky) */
+  int32_t filtered_measurement_vector[INDI_DOF];
+  int32_vect_copy(filtered_measurement_vector, c->filtered_measurement[INDI_NR_FILTERS - 1], INDI_DOF);
+  static int32_t previous_filt_yawrate = 0;
+  filtered_measurement_vector[INDI_YAW] = 512 * (c->filtered_measurement[INDI_NR_FILTERS - 1][INDI_YAW] -
+                                          previous_filt_yawrate);  // = approximately yaw acceleration error
+  previous_filt_yawrate = c->filtered_measurement[INDI_NR_FILTERS - 1][INDI_YAW];
+
+  /* Obtain virtual control input with P controller on pitch and roll */
+  int32_t roll_virtual_control  = (heli_indi_gains.roll_p * att_err.qx)  / 4;
+  int32_t pitch_virtual_control = (heli_indi_gains.pitch_p * att_err.qy) / 4;
+
+  /* Yaw with cascaded PD-controller to generate virtual control */
+  int32_t yaw_rate_reference = (heli_indi_gains.yaw_p * att_err.qz / 8);
+  int32_t yaw_virtual_control = heli_indi_gains.yaw_d * (yaw_rate_reference - body_rate->r);
+
+  /* Set INDI references */
+  c->reference[INDI_ROLL]   = roll_virtual_control;
+  c->reference[INDI_PITCH]  = pitch_virtual_control;
+  c->reference[INDI_YAW]    = yaw_virtual_control;
+  //c->reference[INDI_THRUST] = accel_z_sp;
+
+  /* Subtract (filtered) measurement from reference to get the error */
+  int32_vect_diff(c->error, c->reference, filtered_measurement_vector, INDI_DOF);
+
+  /* Multiply error with inverse of actuator effectiveness, to get delta u (required increment in input) */
+  int32_mat_mul((int32_t **) c->du, (int32_t **) c->invG, (int32_t **) c->error, INDI_DOF, INDI_DOF, 1);
+
+  /* Bitshift back */
+  c->du[INDI_ROLL]  >>= INT32_INVG_FRAC;
+  c->du[INDI_PITCH] >>= INT32_INVG_FRAC;
+  c->du[INDI_YAW]   >>= INT32_INVG_FRAC;
+  c->du[INDI_THRUST] >>= INT32_INVG_FRAC;
+
+  /* Take the current (filtered) actuator position and add the incremental value. */
+  int32_vect_sum(c->u_setpoint, c->filtered_actuator[INDI_NR_FILTERS - 1], c->du, INDI_DOF);
+  //c->u_setpoint[INDI_THRUST] = stabilization_cmd[COMMAND_THRUST];
+
+  /* bound the result */
+  BoundAbs(c->u_setpoint[INDI_ROLL], MAX_PPRZ);
+  BoundAbs(c->u_setpoint[INDI_PITCH], MAX_PPRZ);
+  Bound(c->u_setpoint[INDI_YAW], 0, MAX_PPRZ);
+  Bound(c->u_setpoint[INDI_THRUST], 0.15 * MAX_PPRZ, MAX_PPRZ);
+
+  /* Apply a compensator to the actuator setpoint to obtain actuator command */
+  c->apply_compensator_filters(c->command_out[__k], c->u_setpoint);
+
+  /* At the end, set 'previous' output to current output */
+  int32_vect_copy(c->command_out[__k - 1], c->command_out[__k], INDI_DOF);
+
+  /* Two correction angles, don't rotate but just add.
```

Comment:
The identification procedure gives me pitch_cmd -> pitch_rate and roll_cmd -> roll_rate. These values go directly into the matrix on the diagonal.
Rotating would reduce the control input with a cosine, but I don't want to reduce the control input in these axes. I just need to give some compensation in the other axis proportional to the first.

I agree that the compensations can also be included in the matrix (off diagonal) but these values then would have to be a function of the values on the diagonal to get the same result. I like to have it simply proportional to the command.

The reason I left the matrix in there is:
- If someone continues the project and investigates the couplings better, it might be useful
- If the controller is extended with vertical acceleration control, the effect of collective on yaw acceleration can be easily included.